vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/AbstractLibjnProjectScalarEvaluator.java

```diff
@@ -0,0 +1,170 @@
+    protected final ObjectPointable op;
+    protected TaggedValuePointable key;
+    protected final UTF8StringPointable stringKey;
+    protected final ObjectBuilder ob;
+    protected final ArrayBackedValueStorage abvs, abvs1;
+    protected final SequenceBuilder sb;
+
+    public AbstractLibjnProjectScalarEvaluator(IHyracksTaskContext ctx, IScalarEvaluator[] args) {
+        super(args);
+        this.ctx = ctx;
+        stringKey = (UTF8StringPointable) UTF8StringPointable.FACTORY.createPointable();
+        ob = new ObjectBuilder();
+        abvs = new ArrayBackedValueStorage();
+        abvs1 = new ArrayBackedValueStorage();
+        sb = new SequenceBuilder();
+        op = (ObjectPointable) ObjectPointable.FACTORY.createPointable();
+    }
+
+    @Override
+    protected void evaluate(TaggedValuePointable[] args, IPointable result) throws SystemException {
+        TaggedValuePointable sequence = args[0];
+        TaggedValuePointable keys = args[1];
+        if (keys.getTag() != ValueTag.SEQUENCE_TAG && keys.getTag() != ValueTag.XS_STRING_TAG) {
+            throw new SystemException(ErrorCode.FORG0006);
+        }
+        TaggedValuePointable tempTvp = ppool.takeOne(TaggedValuePointable.class);
+        SequencePointable sp = ppool.takeOne(SequencePointable.class);
+        try {
+            abvs1.reset();
+            sb.reset(abvs1);
+            if (sequence.getTag() == ValueTag.SEQUENCE_TAG) {
+                sequence.getValue(sp);
+                for (int i = 0; i < sp.getEntryCount(); ++i) {
+                    sp.getEntry(i, tempTvp);
+                    if (tempTvp.getTag() == ValueTag.OBJECT_TAG) {
+                        tempTvp.getValue(op);
+                        op.getKeys(tempTvp);
+                        addPairs(tempTvp, keys);
+                    } else {
+                        sb.addItem(tempTvp);
+                    }
+                }
+            } else if (sequence.getTag() == ValueTag.OBJECT_TAG) {
+                sequence.getValue(op);
+                addPairs(tempTvp, keys);
+            } else {
+                sb.addItem(sequence);
+            }
+            sb.finish();
+            result.set(abvs1);
+        } catch (IOException e) {
+            throw new SystemException(ErrorCode.SYSE0001, e);
+        } finally {
+            ppool.giveBack(tempTvp);
+            ppool.giveBack(sp);
+        }
+    }
+
+    protected void addPair(TaggedValuePointable tempTvp, TaggedValuePointable tempValue)
+            throws IOException, SystemException {
+        tempTvp.getValue(stringKey);
+        op.getValue(stringKey, tempValue);
+        ob.addItem(stringKey, tempValue);
+    }
+
+    private void addPairs(TaggedValuePointable objTvp, TaggedValuePointable keys) throws IOException, SystemException {
+        SequencePointable sp1 = ppool.takeOne(SequencePointable.class);
+        TaggedValuePointable tempValue = ppool.takeOne(TaggedValuePointable.class);
```

Comment:
Is "addPairs" nested? If not you can make tempValue a single object that is reused. Similar to stringKey.